services/repository/contributors_graph.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	repo_model "code.gitea.io/gitea/models/repo"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	api "code.gitea.io/gitea/modules/structs"
+	util "code.gitea.io/gitea/modules/util"
+)
+
+// CreateWeeks converts list of sundays to list of *api.WeekData
+func CreateWeeks(sundays []int64) []*api.WeekData {
+	var weeks []*api.WeekData
+	for _, week := range sundays {
+		weeks = append(weeks, &api.WeekData{
+			Week:      week,
+			Additions: 0,
+			Deletions: 0,
+			Commits:   0,
+		},
+		)
+	}
+	return weeks
+}
+
+// GetContributorStats returns contributors stats for git commits for given revision or default branch
+func GetContributorStats(ctx context.Context, repo *repo_model.Repository, revision string) (map[string]*api.ContributorData, error) {
```

Comment:
we should have some cache and data creation lock in place!

else this endpoint will be used to DOS gitea !!!

Reply:
1. first check if in cache -> no -> set a lock and generate + signal client that a new request should be made again soon as data is currently processed
2. if data is generated add it into cache
3. if data exist in cache return it

if web client gets data -> all fine
else if it gets a wait and return later (there should be a specific http code for that (need to look)) go in a loop (with timeout?) that do make retrys

Reply:
This would be nice but I don't have neither time nor energy for this. Unless someone implements this, it is probably going to wait for a long time.

Reply:
ok, I'll look into it